### PR TITLE
Fix #4513: DataTable add globalFilterFunction

### LINKED ISSUE #4513: DataTable: custom global filter function
.. allowing to create your own, custom global filter function.

See also: https://stackoverflow.com/questions/54543368/primefaces-datatable-multiple-keyword-search-using-global-filter

### PR DESCRIPTION
**Pull request by jepsar**
Fix #4513

### CODE REVIEW

**Review by melloware — APPROVED** [review]
I like this idea.

File: src/main/java/org/primefaces/component/datatable/DataTableBase.java [hunk 1]
```diff
@@ -670,6 +670,14 @@ public void setRowEditMode(String rowEditMode) {
         getStateHelper().put(PropertyKeys.rowEditMode, rowEditMode);
     }
 
+    public javax.el.MethodExpression getGlobalFilterFunction() {
```
Comment (Rapster) [inline src/main/java/org/primefaces/component/datatable/DataTableBase.java]: Use simple class name and not with packages please. Good fix though, simple and neat :wink:

Comment (jepsar) [reply src/main/java/org/primefaces/component/datatable/DataTableBase.java]: Sorry about that.. I just did some copy pasting here.

https://github.com/primefaces/primefaces/blob/master/src/main/java/org/primefaces/component/column/ColumnBase.java#L265

File: src/main/java/org/primefaces/component/datatable/feature/FilterFeature.java [hunk 2]
```diff
@@ -180,10 +178,18 @@ public void filter(FacesContext context, DataTable table, List<FilterMeta> filte
...
+        MethodExpression globalFilterFunction = table.getGlobalFilterFunction();
         ELContext elContext = context.getELContext();
 
         for (int i = 0; i < table.getRowCount(); i++) {
             table.setRowIndex(i);
+
+            if (globalFilterFunction != null
+                    && (Boolean) globalFilterFunction.invoke(elContext, new Object[]{table.getRowData(), globalFilterValue, filterLocale})) {
+                filteredData.add(table.getRowData());
+                continue;
```
Comment (jepsar) [inline src/main/java/org/primefaces/component/datatable/feature/FilterFeature.java]: This has kept me thinking.. I think it's wrong. Like this you cannot combine the global filter with column filters. Let me revise this.

**Comment by jepsar** [comment]
@melloware @Rapster [I've revised `FilterFeature.java`](https://github.com/primefaces/primefaces/pull/4522/commits/3c2cd9211beaf94753b85e6efe55c1f33319857e) as mentioned in [my review comment](https://github.com/primefaces/primefaces/pull/4522#pullrequestreview-203037035). Can you please have a look?

Also, I tried to get rid of the conflicts for this branch by [copying changes from the master to `DataTableBase.java`](https://github.com/primefaces/primefaces/pull/4522/commits/89d3a72c1939e845bbf7bf38a7f0d6eae0fa9604), but that did not work... The "Resolve conflicts" button is grey for me, so I wouldn't know what else I could do here.

**Comment by tandraschko** [comment]
can you please fix conflicts?

**Comment by jepsar** [comment]
@tandraschko I wasn't aware at that time that I should create a branch within a fork.. that got me kind of stuck. Shall I provide a new PR without conflicts instead?

**Comment by melloware** [comment]
Yeah if that is easier for you just close this PR and do your branch in your fork.  I typically name my branch like PF4513 or whatever the ticket number is so I can keep my branches straight!

**Comment by jepsar** [comment]
@melloware Okay, I'll provide a new PR. By that way, I'll also provide a second one for PFE TimeAgo somewhere this weekend.

**Comment by jepsar** [comment]
@tandraschko Please use PR #4674 instead of this one

**Pull request closed without merging**